### PR TITLE
[Auth] Add integration tests (headless & webdriver) for middleware

### PR DESCRIPTION
**Pull request by sam-gc**

### CODE REVIEW

File: packages/auth/test/integration/flows/middleware_test_generator.ts [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+import sinonChai from 'sinon-chai';
+
+// eslint-disable-next-line import/no-extraneous-dependencies
+import {Auth, createUserWithEmailAndPassword, User} from '@firebase/auth';
+import { randomEmail } from '../../helpers/integration/helpers';
+
+use(chaiAsPromised);
+use(sinonChai);
+
+export function generateMiddlewareTests(authGetter: () => Auth, signIn: () => Promise<unknown>): void {
```
Comment (sam-gc) [inline packages/auth/test/integration/flows/middleware_test_generator.ts:30]: @jamesdaniels I tried to be exhaustive with these tests. Please take a look and see if I'm missing any big use case.

Cross-tab testing is done down in the webdriver/* files.

I'll add the onAbort callback to the tests once that's merged

File: packages/auth/test/integration/webdriver/redirect.test.ts [hunk 2]
```diff
@@ -70,6 +75,26 @@ browserDescribe('WebDriver redirect IdP test', driver => {
     expect(await driver.call(RedirectFunction.REDIRECT_RESULT)).to.be.null;
   });
 
+  // Redirect works with middleware for now
+  xit('is blocked by middleware', async () => {
```
Comment (hsubox76) [inline packages/auth/test/integration/webdriver/redirect.test.ts:79]: Is this temporarily being skipped because it's not implemented yet?

**Review by hsubox76 — COMMENTED** [review]
Looks good, seems pretty thorough to me.

File: packages/auth/test/integration/webdriver/redirect.test.ts (hunk 2 above)
Comment (sam-gc) [reply packages/auth/test/integration/webdriver/redirect.test.ts:79]: Yep, I'll switch it to `it` once that part is working

**Review by hsubox76 — APPROVED** [review]

**Comment by sam-gc** [comment]
I'm gonna go ahead and submit this. @jamesdaniels if you see anything I need to add later give me a heads up after

**Comment by jamesdaniels** [comment]
@sam-gc SGTM, I'll do some testing next week.

**Pull request merged**